### PR TITLE
Class-based time integration for depletion

### PR DESCRIPTION
**Pull request by drewejohnson**
New abstract class `Integrator` and concrete `CECMIntegrator` for performing depletion analysis. Concrete classes only have to implement the `__call__` method responsible for performing the time integration across a time interval. The abstract base class is responsible for iterating through all time steps, collecting and writing results to file, and executing intermediate transport solutions.

This change was introduced for the following reasons.
1) There is a lot of repeated code across the various integrator functions. This has been moved onto the abstract class
2) This abstraction makes it [or intends to make it] very easy for people to extend the depletion capabilities with custom time integration methods
3) When each integrator only has to implement the integration, not necessarily the normalization and saving, we could add features like specifying time in burnup, and other integration schemes will inherit these features by default.

I chose the `CE/CM` integrator because the predictor is the only one that saves data at the top of the inner loops and the `CECM` has some additional transport solutions that other routines don't have. Unit tests are unchanged thanks to the consistency of the API. 

This is currently a WIP for discussion and API changes before looking at the more complicated schemes.

### CODE REVIEW

File: openmc/deplete/integrator/abc.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        Abstracted to support the predictor's unique save location
+        """
+        Results.save(
+            self.operator, conc_list, results_list, time_list,
+            power, index, proc_time)
+
+    @classmethod
+    def integrate(cls, operator, timesteps, power=None, power_density=None):
+        """High-level interface for depleting with this integrator"""
+        return cls(operator, timesteps, power, power_density).integrate_all()
```
Comment (paulromano) [inline openmc/deplete/integrator/abc.py]: To me, `SomeIntegrator.integate(...)` is not really any better than `SomeIntegrator(...).integrate()`. Consider just removing this classmethod and then renaming `integrate_all` -> `integrate`

**Comment by drewejohnson** [comment]
Thanks for the feedback @paulromano. I left the current functions in place to support any existing depletion scripts, but I definitely agree that a plain class-based approach would be ideal. No class methods either. Just
```
>>> p = Predictor(operator, time, power)
>>> p.integrate()
```

For existing depletion scripts that may use `openmc.deplete.<integrator>`, should we put a `FutureWarning` and then redirect to the class approach?

File: openmc/deplete/integrator/abc.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+class Integrator(ABC):
+    """Abstract class for solving the time-integration for depletion
+
+    """
+
+    def __init__(self, operator, timesteps, power=None, power_density=None):
+        """
+        Parameters
+        ----------
```
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: Will be modified for the full class-approach. I left this here to reconstruct the [`cecm` function docstring](https://github.com/openmc-dev/openmc/pull/1287/commits/cd444510cf701f87c5015b2c1124162c1f0a751a#diff-dd7b8ef0c95cc14521633401517265bbR71).

**Comment by drewejohnson** [comment]
I currently have left the `__init__` docstrings in place on the base class with more mathematical documentation in the class and `__call__` docstrings. This was intentionally done for the following reasons:

1. Introspection by the python `help` function, e.g. `help(CECMIntegrator)` will search back to find the signature and documentation for the parent `__init__` method. This works in IPython and Jupyter environments as well
2. With the exception of the `PredictorIntegrator`, all the integrators fall back to the parent `__init__` directly. Adding the same `Parameters` block to each of the integrators adds a fair amount of code that has to be tracked down and updated if we change how the classes are initialized, e.g. adding burnup steps rather than power.
3. We can tell `sphinx` to render the documentation for the inherited `__init__` method by adding
```
   :special-members: __init__
```
to the template. I've tested this in an [off-shoot branch](https://github.com/drewejohnson/openmc/commit/77369337e512e2db8bcdb1b4ea0a0d38189d0efc) and it works very well.

I can add the `Parameters` documentation to each of the Integrators, but I felt this was a good way to reduce repeated code and (hopefully) reduce pains with future API changes.

**Comment by drewejohnson** [comment]
The class-based integrators use the C API to write statepoint files at the beginning of each depletion step, and at the final transport simulation - 2ee14a4a0f1bfb12302d1b5a03127f4f6d1bf08e.
The files are currently named `"openmc_simulation_n<N>.h5"` but I am open to recommendations by @openmc-dev/technical-committee 

The downside to this is that now each unit test produces a lot of 
```
 Creating state point openmc_simulation_n0.h5
 Creating state point openmc_simulation_n1.h5
...
```
as each of the unit tests makes the call to `openmc.capi.statepoint_write`. Not sure how to turn this off or suppress it during testing that won't adversely effect using this in actual simulation.

File: openmc/deplete/results.py [hunk 3]
```diff
@@ -444,9 +435,9 @@ def save(op, x, op_results, t, power, step_ind, proc_time=None):
 
         for i in range(stages):
             for mat_i in range(n_mat):
-                results[i, mat_i, :] = x[offset + i][mat_i][:]
+                results[i, mat_i, :] = x[i][mat_i]
 
-        results.k = [r.k for r in op_results]
+        results.k = [(r.k.nominal_value, r.k.std_dev) for r in op_results]
```
Comment (paulromano) [inline openmc/deplete/results.py]: Why this change?

File: tests/unit_tests/test_deplete_restart.py [hunk 4]
```diff
@@ -148,25 +139,9 @@ def test_restart_cecm_predictor(run_in_tmpdir):
...
-
-    # Test solution
-    s1 = [1.86872629872102, 1.395525772416039]
-    s2 = [3.32776806576, 2.391425905]
-
-    assert y1[1] == approx(s1[0])
-    assert y2[1] == approx(s1[1])
-
-    assert y1[2] == approx(s2[0])
-    assert y2[2] == approx(s2[1])
```
Comment (paulromano) [inline tests/unit_tests/test_deplete_restart.py]: Are we getting these checks elsewhere?

File: openmc/deplete/integrator/abc.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        operator : openmc.deplete.TransportOperator
```
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: This was the case for previous depletion functions as well
https://github.com/openmc-dev/openmc/blob/f633e848226d6eb8f3dad1677c6c2afd1a7a5e04/openmc/deplete/integrator/cf4.py#L88-L93

Should we add this as an attribute for the `TransportOperator` as well then? Kind of related to the need to add the `power` to `TransportOperator.__call__`

File: openmc/deplete/integrator/abc.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    def __iter__(self):
+        return zip(self.timesteps, self.power)
+
+    def __len__(self):
+        return len(self.timesteps)
+
+    def _get_bos_data_from_openmc(self, step_index, step_power, bos_conc):
+        x = deepcopy(bos_conc)
+        res = self.operator(x, step_power)
+        self._write_statepoint(step_index)
```
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: Agree. My hesitation with `Operator.__call__` is that some routines call that a lot during the integration. Maybe provide a method that is just `write_bos_data` on the base class, that the `Operator` uses to write statepoint data?

File: openmc/deplete/integrator/abc.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+            # Final simulation
+            res_list = [self.operator(conc, p)]
+            self._save_results(
+                [conc], res_list, [t, t], p, self._ires + len(self))
+            self._write_statepoint(len(self))
+
+    def _save_results(self, conc_list, results_list, time_list, power,
+                      index, proc_time=None):
+        """Save the results at the end of of one step"""
+        Results.save(
```
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: I think this was here back when the `predictor` had a unique save location. Will remove

File: tests/unit_tests/test_deplete_restart.py (hunk 4 above)
Comment (drewejohnson) [reply tests/unit_tests/test_deplete_restart.py]: No, because this is a mixed-integration scheme. This test and the `test_restart_predictor_cecm` use two different integrators during the restart [`cecm` then `predictor` and the reverse]. Instead, these tests ensure that one cannot perform these actions.

Relevant commit: a02feed171d9e59c908fab4bc725d0fd3d6dfedd

File: openmc/deplete/results.py (hunk 3 above)
Comment (drewejohnson) [reply openmc/deplete/results.py]: Commit 98957fb47350d8b012236f00cc89a030c38e4590 in this PR changed how the operator stores the k values. Previously, this was a tuple `[k, unc]`. However, the SIE methods update `k` through the iterations, causing issues attempting to multiply / add tuples of floats.

Converting to `ufloats` allowed the k update to be unchanged, but broke the results saving. This specific change was added in that commit to allow the `Results` object to be able to pass the `OperatorResult.k` `ufloat` into what `h5py` can save during:
https://github.com/openmc-dev/openmc/blob/f633e848226d6eb8f3dad1677c6c2afd1a7a5e04/openmc/deplete/results.py#L325-L326

File: openmc/deplete/integrator/abc.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
+from copy import deepcopy
```
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: Sure thing. Any interest in moving the whole integrator module directly into `deplete`? The files are all pretty small now.

File: docs/source/pythonapi/deplete.rst [hunk 9]
```diff
@@ -12,18 +12,18 @@ Josey's thesis, `Development and analysis of high order neutron
...
+    :template:  myclassinherit.rst
+
+    integrator.PredictorIntegrator
+    integrator.CF4Integrator
+    integrator.CECMIntegrator
+    integrator.CELIIntegrator
+    integrator.EPC_RK4_Integrator
+    integrator.LEQIIntegrator
+    integrator.SI_CELI_Integrator
+    integrator.SI_LEQI_Integrator
```
[the remark this answers is not in the document]
Comment (drewejohnson) [reply docs/source/pythonapi/deplete.rst]: Yeah, that was my thought as well :man_shrugging:

File: openmc/deplete/integrator/abc.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: Resolution: 1) method renamed to `_get_bos_data_from_operator`  and 2) added a `write_bos_data` abstract method to `TransportOperator`. This method accepts a single integer of the depletion index, and the `Operator` uses this to write the statepoint data.

This is nice because now our `DummyOperator` used in testing doesn't have to make calls to the C API, as it's `write_bos_data` method can be empty

File: openmc/deplete/integrator/abc.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+            for i, (dt, p) in enumerate(self):
+                if i > 0 or self.operator.prev_res is None:
+                    conc, res = self._get_bos_data_from_openmc(i, p, conc)
+                else:
+                    conc, res = self._get_bos_data_from_restart(i, p, conc)
+                proc_time, conc_list, res_list = self(conc, res.rates, dt, p, i)
+
+                # Insert BOS concentration, transport results
+                conc_list.insert(0, conc)
+                res_list.insert(0, res)
```
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: Two possible alternatives
1) Use slice insertion: `conc_list[0:0] = [conc]`
2) Use a [`deque`](https://docs.python.org/3.7/library/collections.html#collections.deque)

Small timeit comparison:
```
$ python -m timeit -n 100 -r 5 -s "l = list(range(100))" "l.insert(0, 0)"
100 loops, best of 5: 160 nsec per loop

$ python -m timeit -n 100 -r 5 -s "l = list(range(100))" "l[0:0] = [0]"
100 loops, best of 5: 142 nsec per loop

$ python -m timeit -n 100 -r 5 -s "from collections import deque; l = deque(range(100))" "l.insert(0, 0)"
100 loops, best of 5: 90.3 nsec per loop
```

Using a `deque` allows allows us to enforce the number of elements set by each integrator using the `_num_stages` attribute. The list should only be as large as `_num_stages` + 1, to hold intermediate step data and beginning of step data. 

`deque`s are also larger in terms of memory. The largest number of elements these will have to hold, at least for the concrete classes here, is five elements. 

Each concrete `__call__` method would have to create these deques and return them, otherwise we have to create them inside the integrate method. When you include the creation of the `deque` and `list` into the timing comparisons, both insert methods take about 1 usec, while the slice indexing takes 0.9 usec. Sicne we know how many stages to anticipate, we could produce one `conc_list` / `conc_deque` on the `Integrators`, and `__call__` just fills the instance list. Saves us from having to return the concentrations and results lists

:man_shrugging:

File: openmc/deplete/integrator/abc.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: Resolved with 0bc7800a92c8b1f8bc8997219efecf40c39a239f by moving the entire `integrator` sub-module into `openmc.deplete`. Abstract `Integrator` classes live in `openmc.deplete.abc,` with the concrete classes in `openmc.deplete.integrators`.

File: openmc/deplete/integrator/abc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (drewejohnson) [reply openmc/deplete/integrator/abc.py]: Resolved with e286940f8eb3e86ac318fdfe146eec6e23deed9a by passing `prev_res` onto the `TransportOperator` for some basic checking

File: openmc/deplete/integrator/abc.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (paulromano) [reply openmc/deplete/integrator/abc.py]: Eh, let's not worry about it for now. I thought about suggesting `deque` too but like you said, then the `__call__` method would need to return it. Compared to the time to perform a MC calculation, this is going to be way in the noise anyway.

**Comment by paulromano** [comment]
Very nice work @drewejohnson; thanks again!
